src/cycles/simple_cycles.c

```diff
@@ -0,0 +1,459 @@
+ * \brief Search simple cycles starting from one vertex
+ *
+ * \experimental
+ *
+ * \param state The state structure to search on
+ * \param s The vertex index to start search with
+ * \param vertices The vertex IDs of each cycle will be stored here
+ * \param edges The edge IDs of each cycle will be stored here
+ * \return Error code.
+ *
+ * @see https://en.wikipedia.org/wiki/Johnson%27s_algorithm
+ * @see https://stackoverflow.com/a/35922906/3909202
+ * @see https://epubs.siam.org/doi/epdf/10.1137/0204007
+ */
+igraph_error_t igraph_simple_cycles_search_from_one_vertex(
+    igraph_simple_cycle_search_state_t *state, igraph_integer_t s,
+    igraph_vector_int_list_t *vertices, igraph_vector_int_list_t *edges) {
+    // L3:
+    for (igraph_integer_t i = s; i < state->N; ++i) {
+        VECTOR(state->v_blocked)[i] = false;
+        igraph_vector_int_clear(igraph_adjlist_get(&state->B, i));
+    }
+
+    igraph_bool_t found = false;
+    IGRAPH_CHECK(
+        igraph_i_simple_cycles_circuit(state, s, -1, s, vertices, edges, &found));
+
+    for (igraph_integer_t i = 0; i < state->N; ++i) {
+        // We want to remove the vertex with value s, not at position s.
+        // It's fine to use binary search since we never add to, only remove from
+        // an already sorted adjacency list.
+        igraph_integer_t pos;
+        if (igraph_vector_int_binsearch(igraph_adjlist_get(&state->AK, i), s,
+                                        &pos)) {
+            igraph_vector_int_remove(igraph_adjlist_get(&state->AK, i), pos);
+            igraph_vector_int_remove(igraph_inclist_get(&state->IK, i), pos);
+        }
+    }
+    igraph_vector_int_clear(igraph_adjlist_get(&state->AK, s));
+    igraph_vector_int_clear(igraph_inclist_get(&state->IK, s));
+
+    return IGRAPH_SUCCESS;
+}
+
+/**
+ * \function igraph_simple_cycles_search_all
+ * \brief Search all simple cycles
+ *
+ * \experimental
+ *
+ * This function searches for all simple cycles,
+ * using Johnson's cycle detection algorithm
+ * based on the original implementation in:
+ * Johnson DB: Finding all the elementary circuits of a directed graph.
+ * SIAM J Comput 4(1):77-84.
+ * https://epubs.siam.org/doi/10.1137/0204007
+ *
+ *
+ * \param graph The graph to search for
+ * \param vertices The vertex IDs of each cycle will be stored here
+ * \param edges The edge IDs of each cycle will be stored here
+ * \return Error code.
+ */
+igraph_error_t
+igraph_simple_cycles_search_all(const igraph_t *graph,
+                                igraph_vector_int_list_t *v_result,
+                                igraph_vector_int_list_t *e_result) {
+
+    igraph_simple_cycle_search_state_t state;
+
+    IGRAPH_CHECK(igraph_simple_cycle_search_state_init(&state, graph));
+    IGRAPH_FINALLY(igraph_simple_cycle_search_state_destroy, &state);
+
+    // TODO: depending on the graph, it is rather unreasonable to search cycles
+    // from each and every node
```

Comment:
As I understand, duplicates are already taken care of by Johnson's procedure. So avoiding the search from every vertex is probably a minor benefit only.

If I got this right, we need to search from the lowest-index vertex of each cycle. It is not sufficient to search from at least one vertex in each cycle, but we need to search from specifically the lowest index one. Can you confirm that I got this right @GenieTim ? If yes, it seems difficult to reduce the number of starting vertices _efficiently_ (other than skipping those with degree zero and one), and probably not worth bothering. Let me know if you have other ideas.

If we want to optimize more, we can avoid edges that are not in strongly connected components (directed) or biconnected components (undirected), but that's a different optimization, and one that I don't think we need to jump into right away.